src/lab2.py

```diff
@@ -65,55 +65,67 @@ def lex(body):
         breakpoint("lex", text)
     return text
 
-WIDTH, HEIGHT = 800, 600
-HSTEP, VSTEP = 13, 18
-
 SCROLL_STEP = 100
 
-def layout(text):
+def layout(text, width, hstep, vstep):
     display_list = []
-    cursor_x, cursor_y = HSTEP, VSTEP
+    cursor_x, cursor_y = hstep, vstep
     for c in text:
         display_list.append((cursor_x, cursor_y, c))
-        cursor_x += HSTEP
-        if cursor_x >= WIDTH - HSTEP:
-            cursor_y += VSTEP
-            cursor_x = HSTEP
+        cursor_x += hstep
+        if cursor_x >= width - hstep:
+            cursor_y += vstep
+            cursor_x = hstep
         breakpoint("layout", display_list)
     return display_list
 
 class Browser:
-    def __init__(self):
+    def __init__(self, socket, tkinter, width, height, hstep, vstep):
+        self.socket = socket
         self.window = tkinter.Tk()
         self.canvas = tkinter.Canvas(
             self.window,
-            width=WIDTH,
-            height=HEIGHT
-        )
+            width=width,
+            height=height)
+        self.width = width
+        self.height = height
+        self.hstep = hstep
+        self.vstep = vstep
+
         self.canvas.pack()
 
         self.scroll = 0
         self.window.bind("<Down>", self.scrolldown)
 
     def load(self, url):
-        headers, body = request(url)
+        headers, body = request(self.socket, url)
         text = lex(body)
-        self.display_list = layout(text)
+        self.display_list = layout(text, self.width, self.hstep, self.vstep)
         self.draw()
 
     def draw(self):
         self.canvas.delete("all")
         for x, y, c in self.display_list:
             breakpoint("draw")
-            if y > self.scroll + HEIGHT: continue
-            if y + VSTEP < self.scroll: continue
+            if y > self.scroll + self.height: continue
+            if y + self.vstep < self.scroll: continue
             self.canvas.create_text(x, y - self.scroll, text=c)
 
     def scrolldown(self, e):
         self.scroll += SCROLL_STEP
         self.draw()
 
+def my_breakpoint(arg1=None, arg2=None):
```

Comment:
This lab was actually broken before this PR if you tried to run it directly via Python on the commandline,
because the built-in `breakpoint` function does not take any arguments.

Reply:
I also used arg1 and arg2 because the compiler doesn't seem to like the "*args" syntax.

Reply:
Ah, sorry, yeah. You need to run it with `PYTHONBREAKPOINT=0` set, which disables `breakpoint`.